ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java

```diff
@@ -196,6 +197,16 @@ private void handleStatementWithTerminatedQueries(
       executeStatement(
           statement, command, commandId, commandStatusFuture, mode);
     } catch (final KsqlException exception) {
+      if (mode == Mode.RESTORE) {
+        final SqlBaseParser.StatementContext context =
+            ksqlEngine.parse(command.getStatement()).get(0).getStatement().statement();
+        if (context instanceof SqlBaseParser.CreateStreamAsContext
+            || context instanceof SqlBaseParser.CreateTableAsContext
+            || context instanceof SqlBaseParser.InsertIntoContext) {
+          ksqlEngine.incrementQueryId();
```

Comment:
Incrementing the query ID here opens us up to accidentally incrementing the query ID twice, since statement execution could fail after the query ID was incremented through the normal route, and this try-catch wraps the entire `executeStatement` call. For example, here's a check that happens after the query ID is constructed (in `buildPhysicalPlan()` in the line before the one I've linked below) that could fail during recovery due to external state having changed: https://github.com/confluentinc/ksql/blob/6a3a2b9a0f30ebf80e01d959ff375d8b2273bd35/ksql-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java#L143

Reply:
A little hacky, but one option could be to update the `QueryIdGenerator` to take a command topic offset and ensure that the query ID is only incremented at most once for each offset. Since we plan to use command topic offsets in the generation of query IDs moving forward anyway, this hack isn't totally wasted work. It'll just feel a little silly since command topic offsets obviate the need for the `QueryIdGenerator` entirely.